### PR TITLE
Enable CA1822 (Mark members as static) analyzer

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs [hunk 1]
```diff
@@ -93,20 +79,16 @@ public void Export(ICollection<Assembly> assemblies!!)
 
                 Export();
             }
-            catch (Exception ex)
+            catch (Exception)
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs:82]:
```suggestion
            catch
```

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs [hunk 2]
```diff
@@ -119,33 +101,27 @@ public void Export(ICollection<Type> types!!)
...
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs:104]:
```suggestion
            catch
```

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs [hunk 3]
```diff
@@ -119,33 +101,27 @@ public void Export(ICollection<Type> types!!)
...
-            TraceExportBegin();
-
             DataContractSet? oldValue = (_dataContractSet == null) ? null : new DataContractSet(_dataContractSet);
             try
             {
                 AddType(type);
                 Export();
             }
-            catch (Exception ex)
+            catch (Exception)
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs:120]:
```suggestion
            catch
```

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs [hunk 4]
```diff
@@ -263,10 +239,9 @@ public bool CanExport(ICollection<Assembly> assemblies!!)
                 _dataContractSet = oldValue;
                 return false;
             }
-            catch (Exception ex)
+            catch (Exception)
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs:242]:
```suggestion
            catch
```

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs [hunk 5]
```diff
@@ -314,10 +288,9 @@ public bool CanExport(Type type!!)
...
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XsdDataContractExporter.cs:291]:
```suggestion
            catch
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 6]
```diff
@@ -312,7 +312,6 @@ private static string GenerateAssemblyId(Type type)
         internal static bool GenerateSerializerToStream(XmlMapping[] xmlMappings, Type?[] types, string? defaultNamespace, Assembly? assembly, Hashtable assemblies, Stream stream)
         {
             var compiler = new Compiler();
-            try
             {
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs:315]: Why leave the scope?

File: src/coreclr/nativeaot/Runtime.Base/src/System/AttributeUsageAttribute.cs [hunk 7]
```diff
@@ -29,7 +29,7 @@ public AttributeUsageAttribute(AttributeTargets validOn, bool allowMultiple, boo
         // Allowing the set properties as it allows a more readable syntax in the specifiers (and are commonly used)
         // The get properties will be needed only if these attributes are used at Runtime, however, the compiler
         // is getting an internal error if the gets are not defined.
-
+#pragma warning disable CA1822
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/Runtime.Base/src/System/AttributeUsageAttribute.cs:32]: Runtime.Base is not System.Private.CoreLib. it's the runtime library where lowest levels of the runtime are implemented. It has its own System.Object. We might be able to get rid of some of this, but maybe not in this pull request. AttributeUsage doesn't sound like something necessary unless it's forced by Roslyn. Some duplication is necessary because we don't want Runtime.Base to become CoreLib. Some of the duplication is just historic artifacts.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.CoreRT.cs [hunk 8]
```diff
@@ -1407,6 +1403,10 @@ void IEnumerator.Reset()
             {
                 _index = -1;
             }
+
+            public void Dispose()
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.CoreRT.cs:1407]: This is a size optimization. Moving the Dispose method to the generic type means that the method body will be duplicated for each array type. If it's in the non-generic base type, there's only one body. Why is this change needed?

Comment (marek-safar) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.CoreRT.cs]: It was a bug in the analyzer (reverted the change)

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs [hunk 9]
```diff
@@ -174,11 +174,13 @@ public void SetCapacity(HeapIndex heap, int byteCount)
             }
         }
 
+#pragma warning disable CA1822
         // internal for testing
         internal int SerializeHandle(ImmutableArray<int> map, StringHandle handle) => map[handle.GetWriterVirtualIndex()];
         internal int SerializeHandle(BlobHandle handle) => handle.GetHeapOffset();
         internal int SerializeHandle(GuidHandle handle) => handle.Index;
         internal int SerializeHandle(UserStringHandle handle) => handle.GetHeapOffset();
+#pragma warning restore CA1822
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: #done

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs [hunk 10]
```diff
@@ -474,14 +474,15 @@ private DynamicDebugViewEmptyException(SerializationInfo info, StreamingContext
                 : base(info, context)
             {
             }
-
+#pragma warning disable CA1822
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: #removed

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.FileSystem.cs [hunk 11]
```diff
@@ -8,32 +8,32 @@ namespace Microsoft.Win32
 {
     public sealed partial class RegistryKey : MarshalByRefObject, IDisposable
     {
-        private void ClosePerfDataKey()
+        private static void ClosePerfDataKey()
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.FileSystem.cs:11]: Why should this not compile? Both implementations are static and called without instance.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: Actually, I wonder if this was there for the experience in VS, in order to provide some kind of message to the user when viewing something.  @cston, do you know?

[the remark this answers is not in the document]
Comment (cston) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: > I wonder if this was there for the experience in VS, in order to provide some kind of message to the user when viewing something. 

Yes, the EE displays this message as the Value field when a dynamic object has no members. The property is required.

File: src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs [hunk 12]
```diff
@@ -184,7 +184,7 @@ public override bool IsInvalid
         }
 
 #if DEBUG
-        public new IntPtr DangerousGetHandle()
+        public static new IntPtr DangerousGetHandle()
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs]: #removed

File: src/libraries/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs [hunk 13]
```diff
@@ -47,7 +47,7 @@ public override string ToString()
 
 #if DEBUG
         //This method should never be called for this type
-        public new IntPtr DangerousGetHandle()
+        public static new IntPtr DangerousGetHandle()
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: #removed

File: src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs [hunk 14]
```diff
@@ -1424,7 +1417,6 @@ public override void WriteArray(string? prefix, string localName, string? namesp
                 {
                     WriteStartArray(prefix, localName, namespaceUri, count);
                     _writer.WriteDateTimeArray(array, offset, count);
-                    WriteEndArray();
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs:1427]: ok, converted to comment

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: Put it back as there is also explicit reflection test

[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: We usually have comments when a seemingly unnecessary thing exists for tooling reasons.

File: src/libraries/Common/src/System/IO/TempFileCollection.cs [hunk 15]
```diff
@@ -158,7 +158,7 @@ public void Delete()
             SafeDelete();
         }
 
-        internal void Delete(string fileName)
+        internal static void Delete(string fileName)
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Common/src/System/IO/TempFileCollection.cs]: #removed

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (cston) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: > We usually have comments when a seemingly unnecessary thing exists for tooling reasons.

Added in https://github.com/dotnet/runtime/pull/66510.

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs [hunk 16]
```diff
@@ -226,9 +226,10 @@ internal static string GetPipePath(string serverName, string pipeName)
             return s_pipePrefix + pipeName;
         }
 
+#pragma warning disable CA1822
         /// <summary>Throws an exception if the supplied handle does not represent a valid pipe.</summary>
         /// <param name="safePipeHandle">The handle to validate.</param>
-        internal void ValidateHandleIsPipe(SafePipeHandle safePipeHandle)
+        private protected void ValidateHandleIsPipe(SafePipeHandle safePipeHandle)
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs:232]: Why did the visibility here change?  I'm ok with it, I just don't see the connection to the rest of the PR.

File: src/libraries/System.Linq/src/System/Linq/DebugView.cs [hunk 17]
```diff
@@ -52,7 +52,9 @@ public T[] Items
 
     internal sealed class SystemCore_EnumerableDebugViewEmptyException : Exception
     {
+#pragma warning disable CA1822
```
Comment (stephentoub) [inline src/libraries/System.Linq/src/System/Linq/DebugView.cs:55]: Can we add a comment about why this one is needed (similar to the other case)?

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatWriterGenerator.cs [hunk 18]
```diff
@@ -317,7 +317,6 @@ private LocalBuilder LoadMemberValue(DataMember member)
             private void WriteCollection(CollectionDataContract collectionContract)
             {
                 LocalBuilder itemName = _ilg.DeclareLocal(typeof(XmlDictionaryString), "itemName");
-                _ilg.Load(_contextArg);
```
Comment (stephentoub) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatWriterGenerator.cs:320]: What's the purpose of this change?

Comment (stephentoub) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatWriterGenerator.cs:320]: Oh, the member being loaded below changed to a static?

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs [hunk 19]
```diff
@@ -710,15 +710,15 @@ private void OnIndexedHeaderName(int index)
             _state = State.HeaderValueLength;
         }
 
-        private void OnIndexedHeaderNamePostBase(int index)
+        private static void OnIndexedHeaderNamePostBase(int index)
```
Comment (Tratcher) [inline src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs:713]: These won't be static once they're actually implemented.

File: src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs [hunk 20]
```diff
@@ -43,8 +43,6 @@ internal BinaryFormatterWriter(Stream outputStream, ObjectWriter objectWriter, F
             _dataWriter = new BinaryWriter(outputStream, Encoding.UTF8);
         }
 
-        internal void WriteBegin() { }
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs:46]: @marek-safar, please revert all changes to this assembly and just suppress the analyzer for the project.

File: src/libraries/System.Transactions.Local/src/System/Transactions/Configuration/DefaultSettingsSection.cs [hunk 21]
```diff
@@ -10,9 +10,9 @@ internal sealed class DefaultSettingsSection // ConfigurationSection
 
         internal static DefaultSettingsSection GetSection() => s_section;
 
-        public string DistributedTransactionManagerName { get; set; } = ConfigurationStrings.DefaultDistributedTransactionManagerName;
+        public static string DistributedTransactionManagerName { get; set; } = ConfigurationStrings.DefaultDistributedTransactionManagerName;
```
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/src/System/Transactions/Configuration/DefaultSettingsSection.cs:13]: @roji should weigh in on whether any of these changes are going to conflict with his work.

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs (hunk 19 above)
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs:713]: I believe this file is mirrored with aspnetcore. It'd be good to validate whether it's going to break anything when that happens.

**Review by stephentoub — APPROVED** [review]
Other than the remaining comments, LGTM. (I don't plan to re-review the 642 files again ;-)

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs (hunk 19 above)
Comment (marek-safar) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs:713]: I think that's fine, one you implemented them you remove static

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs (hunk 16 above)
Comment (marek-safar) [reply src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs:232]: I think it was to help to test static on the method. I will remove the change.

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs (hunk 19 above)
Comment (marek-safar) [reply src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs:713]: I believe this should be fine as it's private and aspnetcore already had them static (they have CS1822 enabled) before sync with runtime removed it.

https://github.com/dotnet/aspnetcore/pull/39453

File: src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs:46]: Done
